ant-networking/src/network_builder.rs

```diff
@@ -283,20 +284,100 @@ impl NetworkBuilder {
             // How many nodes _should_ store data.
             .set_replication_factor(REPLICATION_FACTOR);
 
-        let (network, net_event_recv, driver) =
-            self.build(kad_cfg, None, true, ProtocolSupport::Outbound, false);
+        let (network, net_event_recv, driver) = self.build(kad_cfg, None, true);
 
         (network, net_event_recv, driver)
     }
 
+    /// Creates a new `ReachabilityCheckSwarmDriver` instance to perform reachability checks.
+    pub fn build_reachability_check_swarm(self) -> Result<ReachabilityCheckSwarmDriver> {
+        let identify_protocol_str = IDENTIFY_PROTOCOL_STR
+            .read()
+            .expect("Failed to obtain read lock for IDENTIFY_PROTOCOL_STR")
+            .clone();
+
+        let peer_id = PeerId::from(self.keypair.public());
+        info!(
+            "Self PeerID {peer_id} is represented as kbucket_key {:?}",
+            PrettyPrintKBucketKey(NetworkAddress::from(peer_id).as_kbucket_key())
+        );
+
+        #[cfg(feature = "open-metrics")]
+        let mut metrics_registries = self.metrics_registries.unwrap_or_default();
+
+        // ==== Transport ====
+        #[cfg(feature = "open-metrics")]
+        let transport = transport::build_transport(&self.keypair, &mut metrics_registries);
+        #[cfg(not(feature = "open-metrics"))]
+        let transport = transport::build_transport(&self.keypair);
+        let transport = if !self.local {
+            debug!("Preventing non-global dials");
+            // Wrap upper in a transport that prevents dialing local addresses.
+            libp2p::core::transport::global_only::Transport::new(transport).boxed()
+        } else {
+            transport
+        };
+
+        let (relay_transport, _relay_client) =
+            libp2p::relay::client::new(self.keypair.public().to_peer_id());
+        let transport = relay_transport
+            .upgrade(libp2p::core::upgrade::Version::V1Lazy)
+            .authenticate(
+                libp2p::noise::Config::new(&self.keypair)
+                    .expect("Signing libp2p-noise static DH keypair failed."),
+            )
+            .multiplex(libp2p::yamux::Config::default())
+            .or_transport(transport);
+
+        let transport = transport
+            .map(|either_output, _| match either_output {
+                Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+                Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+            })
+            .boxed();
+
+        // Listen on the provided address
+        let listen_socket_addr = self
+            .listen_addr
+            .ok_or(NetworkError::ListenAddressNotProvided)?;
+
+        // Identify Behaviour
+        let agent_version = IDENTIFY_REACHABILITY_CHECK_CLIENT_VERSION_STR
```

Comment:
This means the reachability only got supported by the new version nodes ?

Wondering why this is required.

This could result in backward compability issue ? i.e. not only old nodes actually blocks or got blocked, but also only after most of nodes got upgraded, can this feature to be fully functional ?

Reply:
> but also only after most of nodes got upgraded, can this feature to be fully functional ?

That is right, older nodes cannot help us in testing out the "reachability" of nodes as they don't have the 180s wait before dialing back. Hence adding another dependency here is fine.